source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,279 @@
+#include "extensions/filters/http/jwt_authn/verifier.h"
+
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtProvider;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirement;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirementAndList;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirementOrList;
+using ::google::jwt_verify::Status;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+
+/**
+ * Struct to keep track of verifier completed and responded state for a request.
+ */
+struct CompletionState {
+  // if verifier node has responded to a request or not.
+  bool is_completed_{false};
+  // number of completed inner verifier for an any/all verifier.
+  std::size_t number_completed_children_{0};
+};
+
+class ContextImpl : public Verifier::Context {
+public:
+  ContextImpl(Http::HeaderMap& headers, Verifier::Callbacks* callback)
+      : headers_(headers), callback_(callback) {}
+
+  Http::HeaderMap& headers() const override { return headers_; }
+
+  Verifier::Callbacks* callback() const override { return callback_; }
+
+  void cancel() override {
+    for (const auto& it : auths_) {
+      it->onDestroy();
+    }
+    auths_.clear();
+  }
+
+  // Get Response data which can be used to check if a verifier node has responded or not.
+  CompletionState& getCompletionState(const Verifier* verifier) {
+    return completion_states_[verifier];
+  }
+
+  // Stores an authenticator object for this request.
+  void storeAuth(AuthenticatorPtr&& auth) { auths_.emplace_back(std::move(auth)); }
+
+private:
+  Http::HeaderMap& headers_;
+  Verifier::Callbacks* callback_;
+  std::unordered_map<const Verifier*, CompletionState> completion_states_;
+  std::vector<AuthenticatorPtr> auths_;
+};
+
+// base verifier for provider_name, provider_and_audiences, and allow_missing_or_failed.
+class BaseVerifierImpl : public Verifier {
+public:
+  BaseVerifierImpl(const BaseVerifierImpl* parent) : parent_(parent) {}
+
+  void onCompleteHelper(Status status, ContextImpl& context) const {
```

Comment:
rename this function to "completeWithStatus" to indicate it is actually done the verification and calling upper layer onCompleted() callback